shared-bindings/crypto/__init__.c

```diff
@@ -0,0 +1,86 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdint.h>
+
+#include "py/obj.h"
+#include "py/runtime.h"
+
+#include "__init__.h"
+
+//| :mod:`Crypto.Cipher` --- Cipher encryption and decryption routines
+//| ==================================================================
+//|
+//| .. module:: Crypto.Cipher
+//|   :synopsis: Embedded implementation of PyCrypto
```

Comment:
We don't want to mimic pycrypto. I got the wrong python module. (PyCrypto was last released in 2013.) It looks like `cryptography` is the recommended package: https://docs.python-guide.org/scenarios/crypto/

Instead, I think we want our own APIs that only feature `encrypt_into` and `decrypt_into` that both take two arguments, the source and destination buffers. They can be the same or different. Omitting the methods that return new memory means folks are less likely to generate a bunch of garbage they don't want.

Reply:
Is there an API reference for that library?  I used to have a simpler api with an `encrypt()` and `decrypt()` call, but I threw that away to create API compatibility with PyCrypto.

What should the API look like when it comes to instantiation, key and IV management, and actual usage?

Reply:
Yup! It is here: https://cryptography.io/en/latest/

It's not super clear to me how you arbitrarily assemble pieces from the `hazmat` layer.